python/ffcc/io.py

```diff
@@ -172,26 +180,39 @@ def _list_subfolders(path):
     return subfolders
 
   folders = [os.path.join(path, f) for f in _list_subfolders(path)]
-  data = []
+  # include the main directory
+  folders.append(path)
+  image_files = []
   for folder in folders:
-    try:
-      illuminant, weight, frames = read_from_directory(folder)
-      for i, f in enumerate(frames):
-        rgb, extended_feature = f
-        data.append({
-            'name': '{}_{}'.format(os.path.basename(folder), i),
-            'rgb': rgb,
-            'extended_feature': extended_feature,
-            'illuminant': illuminant,
-            'weight': weight,
-        })
-    except:  # pylint: disable=bare-except
-      tf.logging.error('Unable to load %s', folder)
-
-  return data
-
-
-def write_dataset_to_tfrecord(dataset, output_dir):
-  """Encodes dataset into TFRecords and write into the destination folder."""
-  for d in dataset:
-    encode_to_tfrecord(d, output_dir)
+    if folder:
+      image_files = [
+        os.path.join(folder, file) for file in os.listdir(folder) if
+        file.endswith('.png')]
+
+  # check if cvfolds.txt file exists in the root path
+  if tf.io.gfile.exists(os.path.join(path, 'cvfolds.txt')):
+    cvfolds = np.loadtxt(os.path.join(path, 'cvfolds.txt'), dtype=np.int)
+  else:
+    # create a 3-fold cross-validation partition and store it in the root path
+    number_of_files = len(image_files)
+    indices = np.linspace(0, number_of_files - 1, number_of_files).astype(int)
+    cvfolds = np.zeros(number_of_files).astype(int)
+    random.shuffle(indices)
+    for fold in range(3):
+      cvfolds[indices[np.floor((fold * number_of_files) / 3).astype(int):
+                      np.floor(((fold + 1) * number_of_files) / 3).astype(int)
+              ]] = int(fold + 1)
+    cvfolds_file = open(os.path.join(path, 'cvfolds.txt'), 'w')
+    np.savetxt(cvfolds_file, cvfolds, fmt='%d')
+    cvfolds_file.seek(cvfolds_file.tell() - 1, os.SEEK_SET)
+    cvfolds_file.truncate()
+    cvfolds_file.close()
+
```

Comment:
What does this sequence of file operations achieve? If you just want to write to the file in append mode, pass 'a' in open?

Also use with: with open(...) as cvfolds_file: ...

Reply:
This is to avoid the extra newline added after the final element. I will use with open(..).

Reply:
We shouldn't need to avoid the extra newline - the parsing code should be robust to newlines. If you read a file with (for line in file: ...), I think this excludes the blank space at the end. Can you point me to the code that reads this data?

Reply:
Honestly, I didn't try it without this handling, so yes it may work without a need for this truncate operation. Here is the loading line:
https://github.com/mahmoudnafifi/ffcc/blob/fc83895bf1c39f1ac3f58a347143caa778195023/python/ffcc/io.py#L196

Reply:
Yes, let's try the simple solution and see if that works.

Reply:
It is handled automatically by the parsing code. Fixed!